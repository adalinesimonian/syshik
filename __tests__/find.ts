import syshik, { find } from '../src/index'

describe('argument types', () => {
  it('find should be a function', () => {
    expect(typeof syshik.find).toBe('function')
    expect(typeof find).toBe('function')
  })

  it('find should throw when called with a non-string argument', () => {
    expect(() => find(1 as any, '')).toThrowErrorMatchingSnapshot()
    expect(() => find(true as any, '')).toThrowErrorMatchingSnapshot()
    expect(() => find(null as any, '')).toThrowErrorMatchingSnapshot()
    expect(() => find(undefined as any, '')).toThrowErrorMatchingSnapshot()
    expect(() => find([] as any, '')).toThrowErrorMatchingSnapshot()
    expect(() => find({} as any, '')).toThrowErrorMatchingSnapshot()
    expect(() => find((() => {}) as any, '')).toThrowErrorMatchingSnapshot()
  })

  it('find should throw when called with a non-string and non-string-array search argument', () => {
    expect(() => find('', 1 as any)).toThrowErrorMatchingSnapshot()
    expect(() => find('', true as any)).toThrowErrorMatchingSnapshot()
    expect(() => find('', null as any)).toThrowErrorMatchingSnapshot()
    expect(() => find('', undefined as any)).toThrowErrorMatchingSnapshot()
    expect(() => find('', {} as any)).toThrowErrorMatchingSnapshot()
    expect(() => find('', (() => {}) as any)).toThrowErrorMatchingSnapshot()
  })
})

describe('return types', () => {
  describe('when called with a string argument', () => {
    it('find should return empty array when no matches', () => {
      expect(find('', 'abc')).toEqual([])
    })

    it('find should return array of matches when there are matches', () => {
      expect(find('abc', 'abc')).toEqual(['abc'])
      expect(find('abcabcabc', 'abc')).toEqual(['abc', 'abc', 'abc'])
    })
  })

  describe('when called with a string array argument', () => {
    it('find should an object with empty arrays when no matches', () => {
      expect(find('', ['abc', 'def'])).toEqual({
        abc: [],
        def: [],
      })
    })

    it('find should return an object of matches when there are matches', () => {
      expect(find('abc', ['abc'])).toEqual({
        abc: ['abc'],
      })
      expect(find('abcdef', ['abc', 'def'])).toEqual({
        abc: ['abc'],
        def: ['def'],
      })
      expect(find('abcdef', ['bcd', 'cde'])).toEqual({
        bcd: ['bcd'],
        cde: ['cde'],
      })
      expect(find('abcdef', ['abc', 'def', 'ghi'])).toEqual({
        abc: ['abc'],
        def: ['def'],
        ghi: [],
      })
    })
  })
})

describe('homoglyphs and lookalikes', () => {
  it('find should match diacritic characters', () => {
    expect(find('what éggplâńt', 'eggplant')).toEqual(['éggplâńt'])
  })

  it('find should match homoglyphs', () => {
    expect(find('𝜧𝚘𝝄𝕟 𐒴ⅰ𝕧ҽʀ', 'moon river')).toEqual(['𝜧𝚘𝝄𝕟 𐒴ⅰ𝕧ҽʀ'])
  })

  it('find should match lookalikes', () => {
    expect(find('|=/-\\|<3', 'fake')).toEqual(['|=/-\\|<3'])
  })
})
