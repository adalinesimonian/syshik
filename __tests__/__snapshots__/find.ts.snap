// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`argument types find should throw when called with a non-string and non-string-array search argument 1`] = `"Search strings must either be passed as a string or an array of strings, found number"`;

exports[`argument types find should throw when called with a non-string and non-string-array search argument 2`] = `"Search strings must either be passed as a string or an array of strings, found boolean"`;

exports[`argument types find should throw when called with a non-string and non-string-array search argument 3`] = `"Search strings must either be passed as a string or an array of strings, found object"`;

exports[`argument types find should throw when called with a non-string and non-string-array search argument 4`] = `"Search strings must either be passed as a string or an array of strings, found undefined"`;

exports[`argument types find should throw when called with a non-string and non-string-array search argument 5`] = `"Search strings must either be passed as a string or an array of strings, found object"`;

exports[`argument types find should throw when called with a non-string and non-string-array search argument 6`] = `"Search strings must either be passed as a string or an array of strings, found function"`;

exports[`argument types find should throw when called with a non-string argument 1`] = `"Expected a string, found number"`;

exports[`argument types find should throw when called with a non-string argument 2`] = `"Expected a string, found boolean"`;

exports[`argument types find should throw when called with a non-string argument 3`] = `"Expected a string, found object"`;

exports[`argument types find should throw when called with a non-string argument 4`] = `"Expected a string, found undefined"`;

exports[`argument types find should throw when called with a non-string argument 5`] = `"Expected a string, found object"`;

exports[`argument types find should throw when called with a non-string argument 6`] = `"Expected a string, found object"`;

exports[`argument types find should throw when called with a non-string argument 7`] = `"Expected a string, found function"`;
